R/shinyui.R

```diff
@@ -77,7 +77,7 @@ renderPage <- function(ui, connection, showcase=0, testMode=FALSE) {
   }
 
   html <- renderDocument(ui, shiny_deps, processDep = createWebDependency)
-  writeUTF8(html, con = connection)
+  enc2utf8(paste(collapse = "\n", html))
```

Comment:
I'm pretty sure `renderDocument` returns a UTF-8 encoded string vector with one element (and class `"html"`), so you should be able to just call `as.character` (to remove class and other attributes) on the value of `renderDocument()`.

https://github.com/rstudio/htmltools/blob/a8025f360/R/template.R#L134-L136

Example:

```R
library(htmltools)
library(shiny)

ui <- htmlTemplate(
  system.file("template", "default.html", package = "shiny"),
  body = fluidPage(div("hello world"))
)

shiny_deps <- list(
  htmlDependency("json2", "2014.02.04", c(href="shared"), script = "json2-min.js"),
  htmlDependency("shiny", utils::packageVersion("shiny"), c(href="shared"),
    script = if (getOption("shiny.minified", TRUE)) "shiny.min.js" else "shiny.js",
    stylesheet = "shiny.css")
)

res <- renderDocument(ui, shiny_deps)
length(res)
#> [1] 1
str(res)
#>  'html' chr "<!DOCTYPE html>\n<html>\n<head>\n  <meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\"/>\n  "| __truncated__
#>  - attr(*, "html")= logi TRUE
```

Reply:
TBH I'd rather keep it the way I have it just to be safe, enc2utf8 and paste should have no perf impact if it's already a UTF-8 encoded single element character.

Reply:
(@wch told me offline it's ok to resolve)